streaming/src/main/scala/org/apache/spark/streaming/dstream/SocketInputDStream.scala

```diff
@@ -52,46 +52,60 @@ class SocketReceiver[T: ClassTag](
     storageLevel: StorageLevel
   ) extends Receiver[T](storageLevel) with Logging {
 
+  private var socket: Socket = _
+
   def onStart() {
-    // Start the thread that receives data over a connection
-    new Thread("Socket Receiver") {
-      setDaemon(true)
-      override def run() { receive() }
-    }.start()
+
+    try {
+      logInfo(s"Connecting to $host:$port")
+      socket = new Socket(host, port)
+      logInfo(s"Connected to $host:$port")
+
+      // Start the thread that receives data over a connection
+      new Thread("Socket Receiver") {
+        setDaemon(true)
+        override def run() { receive() }
+      }.start()
+    } catch {
+      case e: ConnectException =>
```

Comment:
Good question. If it fails to connect, it seems like that should be a fatal error and not attempt to restart. So I suppose you could not catch anything in `onStart` as failures here are more fatal. Later, it makes sense to still catch general exceptions (not `ConnectException`) and maybe try to restart